python/deltalake/writer.py

```diff
@@ -235,8 +239,38 @@ def visitor(written_file: Any) -> None:
         invariants = table.schema().invariants
         checker = _DeltaDataChecker(invariants)
 
+        def check_data_is_aligned_with_partition_filtering(
+            batch: pa.RecordBatch,
+        ) -> None:
+            def arrow_value_to_partition_string(val: Any) -> str:
+                if val.type == "bool":
+                    val = str(val).lower()
+                else:
+                    val = str(val)
+                return val
+
+            if table is None:
+                return
+            allowed_partitions = table._table.get_active_partitions(partitions_filters)
+            for column_index, column_name in enumerate(batch.schema.names):
+                if column_name in table.metadata().partition_columns:
+                    for value in batch.column(column_index).unique():
+                        print(type(value))
+                        if (
+                            column_name,
+                            arrow_value_to_partition_string(value),
+                        ) not in allowed_partitions:
+                            raise ValueError(
+                                f"Data should be aligned with partitioning. "
+                                f"Partition '{column_name}'='{value}' should be filtered out from data."
```

Comment:
Do we allow writing to a *new* partition with this method? It seems like the partition must already exist. We might want to either allow that or provide a more helpful error message if `allowed_partitions` comes back empty.

Reply:
Now it allows create new partition and overwrites existing ones. Not sure if we want to have it configurable.